### PR TITLE
Filter options incompatible with TLS 1.3

### PR DESCRIPTION
**Pull request by lukebakken**
See #313

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -5,9 +5,9 @@ on: [push, pull_request]
 jobs:
   test:
     name: Test
-    runs-on: ubuntu-16.04
+    runs-on: ubuntu-18.04
     strategy:
-      fail-fast: false
```
Comment (ericmj) [inline .github/workflows/main.yml:10]: Please revert this change.

File: test/mint/integration_test.exs [hunk 2]
```diff
@@ -210,4 +210,25 @@ defmodule Mint.IntegrationTest do
       end
     end
   end
+
+  describe "force TLS v1.3 only" do
+    @describetag :integration
+
+    test "rabbitmq.com" do
+      if Mint.Core.Transport.SSL.ssl_version() >= [10, 2] do
+        ciphers = :ssl.filter_cipher_suites(:ssl.cipher_suites(:all, :"tlsv1.3"), [])
```
Comment (ericmj) [inline test/mint/integration_test.exs:219]: Is this necessary?

File: lib/mint/core/transport/ssl.ex [hunk 3]
```diff
@@ -445,6 +446,23 @@ defmodule Mint.Core.Transport.SSL do
...
 
+  defp remove_incompatible_ssl_opts(opts) do
+    # Note: these are the TLS versions that are compatible with :reuse_sessions and :secure_renegotiate
+    # If none of the compatible TLS versions are present in transport_opts, then :reuse_sessions and
+    # :secure_renegotiate will be removed from transport_opts
+    tls_compatible_set = MapSet.new([:tlsv1, :"tlsv1.1", :"tlsv1.2"])
+
+    with tls_versions_opt when not is_nil(tls_versions_opt) <- Keyword.get(opts, :versions),
+         tls_versions_set <- MapSet.new(tls_versions_opt),
+         0 <- MapSet.size(MapSet.intersection(tls_compatible_set, tls_versions_set)) do
```
Comment (ericmj) [inline lib/mint/core/transport/ssl.ex:457]: This can be simplified by doing `if :"tlsv1.3" in Keyword.get(opts, :versions, [])`.

File: test/mint/integration_test.exs (hunk 2 above)
Comment (lukebakken) [reply test/mint/integration_test.exs:219]: Yes, for the same reason I had to add [this code](https://github.com/elixir-mint/mint/blob/8e6dab5d6c4b57097e2128d845afd58f76c9018a/test/mint/http2/integration_test.exs#L12-L15) in #310

File: lib/mint/core/transport/ssl.ex (hunk 3 above)
Comment (lukebakken) [reply lib/mint/core/transport/ssl.ex:457]: Actually, that isn't equivalent. If a user specifies `{:versions, [:"tlsv1.2", :"tlsv1.3"]}` then it's still OK to specify `:reuse_sessions` and `:secure_renegotiate` according to the validation done in `ssl.erl`.

Comment (lukebakken) [reply lib/mint/core/transport/ssl.ex:457]: At this point in time we know that the set of valid TLS versions to use `:reuse_sessions` and `:secure_renegotiate` is in `tls_compatible_set`. By hard-coding `:"tlsv1.3"` we'll probably have to re-visit this code in the future. My guess is that TLS 1.4 and beyond also will not be compatible with those two settings.

Comment (ericmj) [reply lib/mint/core/transport/ssl.ex:457]: AFAIK TLS 1.4 development hasn't been announced so it would be many years until it was implemented in OTP and we would likely have to revisit the code around it when it's added anyway so I would rather keep it simple now.

File: test/mint/integration_test.exs (hunk 2 above)
Comment (ericmj) [reply test/mint/integration_test.exs:219]: Since this PR is about automatically handle some options for TLS 1.3 would it be appropriate to add the correct ciphers automatically as well?

Comment (lukebakken) [reply test/mint/integration_test.exs:219]: Personally I would not want a library setting ciphers for me. At any rate, that feature should be in a different PR.

File: lib/mint/core/transport/ssl.ex (hunk 3 above)
Comment (lukebakken) [reply lib/mint/core/transport/ssl.ex:457]: This is a valid set of options that would be modified by your suggestion:

```
{:versions, [:"tlsv1.2", :"tlsv1.3"]}
```

The current code mirrors the validation done in `ssl.erl`

Comment (lukebakken) [reply lib/mint/core/transport/ssl.ex:457]: Oh wait I think I misunderstood your suggestion. I'll get it later. FYI you do have permissions to modify my branch & push to my fork :-)

File: test/mint/integration_test.exs (hunk 2 above)
Comment (ericmj) [reply test/mint/integration_test.exs:219]: We already set default ciphers, so it would only be a change to the defaults for TLS 1.3. Agreed that it fits better in an a separate PR.

Comment (lukebakken) [reply test/mint/integration_test.exs:219]: https://github.com/elixir-mint/mint/issues/316

**Comment by lukebakken** [comment]
@ericmj all set for re-review. Thanks.

File: lib/mint/core/transport/ssl.ex [hunk 4]
```diff
@@ -445,6 +446,28 @@ defmodule Mint.Core.Transport.SSL do
...
+    # Note: these are the TLS versions that are compatible with :reuse_sessions and :secure_renegotiate
+    # If none of the compatible TLS versions are present in transport_opts, then :reuse_sessions and
+    # :secure_renegotiate will be removed from transport_opts
+    tls_compatible_set = MapSet.new([:tlsv1, :"tlsv1.1", :"tlsv1.2"])
+
+    tls_versions_opt = Keyword.get(opts, :versions, [])
+
+    if :"tlsv1.3" in tls_versions_opt do
+      with tls_versions_set <- MapSet.new(tls_versions_opt),
+           0 <- MapSet.size(MapSet.intersection(tls_compatible_set, tls_versions_set)) do
```
Comment (ericmj) [inline lib/mint/core/transport/ssl.ex:459]: Is the `with` necessary when we have the `if` above?

Comment (lukebakken) [reply lib/mint/core/transport/ssl.ex:459]: Yes, for [this reason](https://github.com/elixir-mint/mint/pull/314#discussion_r619848922). The presence of `:"tlsv1.3"` isn't enough, we have to see if _any_ of the compatible-with-`:reuse_sessions` versions are present.

For instance, this will probably be a typical user-specified value at some point:

```
{:versions, [:"tlsv1.2", :"tlsv1.3"]}
```

If the code just looks for `:tlsv1.3` then we would be removing settings when we shouldn't.

Comment (ericmj) [reply lib/mint/core/transport/ssl.ex:459]: In that case let's do `unless Enum.any?(compatible_versions, &(&1 in actual_versions))`

Comment (lukebakken) [reply lib/mint/core/transport/ssl.ex:459]: AHA! Thanks, I figured there was a more idiomatic way but couldn't really figure out what.

**Review by ericmj — APPROVED** [review]

**Comment by lukebakken** [comment]
@ericmj are we waiting on another review? Otherwise, if you have a sec please merge. Going to start on #316

**Comment by ericmj** [comment]
Yes, we want two reviews before merging.

**Comment by lukebakken** [comment]
FWIW, my `finch` PR (https://github.com/keathley/finch/pull/133) depends on this PR and a release of `mint` version `1.3`.

**Review by whatyouhide — APPROVED** [review]
